File: autocontext/src/autocontext/execution/agent_task_evolution.py
"""Multi-generation support for AgentTask scenarios (AC-281)."""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass, field
from typing import Any

from pydantic import BaseModel, Field

from autocontext.knowledge.compaction import compact_prompt_component
from autocontext.scenarios.agent_task import AgentTaskResult


class AgentTaskGenerationState(BaseModel):
    """Cross-generation state for an agent task evolution run."""

    generation: int
    best_output: str
    best_score: float
    playbook: str
    score_history: list[float]
    lesson_history: list[str]
    metadata: dict[str, Any] = Field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        return self.model_dump()

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> AgentTaskGenerationState:
        return cls.model_validate(data)


@dataclass(slots=True)
class LessonSignal:
    """Structured, evaluator-provided guidance for the next generation.

    Domain-aware lesson accumulation: a deterministic evaluator usually
    knows *why* a candidate plateaued and *what* to try next (the size
    delta, which constraints bind, an explicit hint). Carrying that here
    lets ``accumulate_lessons`` write actionable playbook entries instead
    of only score + dimension scores.
    """

    hint: str = ""
    plateau: bool = False
    metrics: dict[str, float] = field(default_factory=dict)


@dataclass(slots=True)
class AgentTaskGenerationEvaluation:
    """Evaluation result for one cross-generation candidate."""

    output: str
    score: float
    reasoning: str
    dimension_scores: dict[str, float] = field(default_factory=dict)
    round_count: int = 1
    met_threshold: bool = False
    metadata: dict[str, Any] = field(default_factory=dict)
    lesson_signal: LessonSignal | None = None


@dataclass(frozen=True, slots=True)
class FunctionSlot:
    """A fixed code harness with a small evolved slot (AC-776).

    Function-slot evolution mode keeps the evolved unit small: the runner
    carries only the slot in state and in the enriched prompt (so prompts
    stay compact), while evaluation runs the assembled ``harness`` + slot.
    This avoids the whole-program-bloat failure mode where carrying a large
    generated artifact in ``best_output`` ballooned every prompt.

    Convention: the slot is *prepended* to the harness, so the harness can
    reference names the slot defines (e.g. a ``priority`` function the
    greedy skeleton calls).
    """

    harness: str

    def assemble(self, slot: str) -> str:
        """Return the full runnable program: slot prepended to harness."""
        return f"{slot}\n\n{self.harness}"


def accumulate_lessons(
    judge_result: AgentTaskResult,
    generation: int,
    signal: LessonSignal | None = None,
) -> str:
    """Extract a structured lesson from judge feedback for the playbook.

    When the evaluator supplies a :class:`LessonSignal`, its actionable
    guidance (hint, plateau flag, metrics) is rendered alongside the score
    and dimension scores so the playbook carries move-level direction.
    """
    parts: list[str] = [f"Generation {generation} (score: {judge_result.score:.2f}):"]

    if judge_result.reasoning:
        parts.append(f"  Feedback: {judge_result.reasoning}")

    weak_dims = {dim: score for dim, score in judge_result.dimension_scores.items() if score < 0.7}
    if weak_dims:
        dim_strs = [f"{dim} ({score:.2f})" for dim, score in sorted(weak_dims.items(), key=lambda x: x[1])]
        parts.append(f"  Weak dimensions: {', '.join(dim_strs)}")

    strong_dims = {dim: score for dim, score in judge_result.dimension_scores.items() if score >= 0.8}
    if strong_dims:
        dim_strs = [f"{dim} ({score:.2f})" for dim, score in sorted(strong_dims.items(), key=lambda x: -x[1])]
        parts.append(f"  Strong dimensions: {', '.join(dim_strs)}")

    if not judge_result.reasoning and not weak_dims:
        parts.append(f"  Score: {judge_result.score:.2f}")

    if signal is not None:
        if signal.hint:
            parts.append(f"  Hint: {signal.hint}")
        if signal.plateau:
            parts.append(
                "  Plateau detected — a structurally different approach is needed; "
                "incremental tweaks are not advancing the score."
            )
        if signal.metrics:
            metric_strs = [f"{k}={v:g}" for k, v in sorted(signal.metrics.items())]
            parts.append(f"  Metrics: {', '.join(metric_strs)}")

    return "\n".join(parts)


def build_enriched_prompt(
    *,
    task_prompt: str,
    playbook: str,
    generation: int,
    best_output: str,
    best_score: float,
    harness: str = "",
) -> str:
    """Enrich a task prompt with cross-generation context.

    In function-slot mode (``harness`` provided), the fixed harness is shown
    once as stable context so the model knows the contract it writes the slot
    against. The evolved slot itself is carried via ``best_output``.
    """
    playbook = compact_prompt_component("agent_task_playbook", playbook)
    best_output = compact_prompt_component("agent_task_best_output", best_output)
    sections: list[str] = [task_prompt]

    if harness:
        sections.append(f"\n\n## Fixed Harness (do not modify; you write only the slot)\n{harness}")

    if playbook:
        sections.append(
            f"\n\n## Accumulated Lessons (Generation {generation})\nPrevious best score: {best_score:.2f}\n\n{playbook}"
        )

    if best_output:
        sections.append(f"\n\n## Best Previous Output (score {best_score:.2f})\n{best_output}")

    if playbook or best_output:
        sections.append(
            "\n\nUse the accumulated lessons and previous best output as context. "
            "Produce an improved version that addresses the identified weaknesses."
        )

    return "\n".join(sections)


def migrate_states(
    states: list[AgentTaskGenerationState],
) -> list[AgentTaskGenerationState]:
    """Island migration: seed lagging islands with the global champion.

    Each island below the best score adopts the champion's best output and
    score (so winners propagate), but keeps its own playbook so accumulated
    lessons stay diverse. The champion island (and any tied) are unchanged.
    """
    if not states:
        return states
    champion = max(states, key=lambda s: s.best_score)
    migrated: list[AgentTaskGenerationState] = []
    for s in states:
        if s.best_score < champion.best_score:
            migrated.append(
                s.model_copy(
                    update={
                        "best_output": champion.best_output,
                        "best_score": champion.best_score,
                    }
                )
            )
        else:
            migrated.append(s)
    return migrated


class AgentTaskTrajectory(BaseModel):
    """Trajectory report for a multi-generation agent task run."""

    task_name: str
    total_generations: int
    score_history: list[float]
    lessons_per_generation: list[int]
    cold_start_score: float
    final_score: float
    improvement_delta: float
    metadata: dict[str, Any] = Field(default_factory=dict)

    def cold_vs_warm_summary(self) -> str:
        """Human-readable comparison of cold-start vs warmed performance."""
        lines = [
            f"Task: {self.task_name}",
            f"Generations: {self.total_generations}",
            f"Cold-start score: {self.cold_start_score:.2f}",
            f"Final score: {self.final_score:.2f}",
            f"Improvement: +{self.improvement_delta:.2f}",
        ]
        if len(self.score_history) >= 2:
            lines.append(f"Trajectory: {' → '.join(f'{score:.2f}' for score in self.score_history)}")
        return "\n".join(lines)

    def to_dict(self) -> dict[str, Any]:
        return self.model_dump()

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> AgentTaskTrajectory:
        return cls.model_validate(data)


class ScenarioFamilyGuide:
    """When-to-use guidance for choosing between scenario families."""

    def __init__(self) -> None:
        self.families: dict[str, dict[str, str]] = {
            "agent_task": {
                "when_to_use": (
                    "Open-ended rubric-driven tasks evaluated by an LLM judge. "
                    "Best for writing, analysis, code review, and other subjective "
                    "tasks where quality is dimension-scored."
                ),
                "multi_gen": "Yes — via AgentTaskEvolutionRunner with playbook carry-forward.",
            },
            "simulation": {
                "when_to_use": (
                    "Richly stateful scenarios with world state, entities, resources, "
                    "and multi-step transitions. Best for orchestration, planning, "
                    "and resource-management tasks."
                ),
                "multi_gen": "Yes — via GenerationRunner with ScenarioInterface.",
            },
            "negotiation": {
                "when_to_use": (
                    "Multi-party interaction scenarios with offers, counteroffers, "
                    "and agreement dynamics. Best for bargaining and diplomacy."
                ),
                "multi_gen": "Yes — via GenerationRunner.",
            },
            "schema_evolution": {
                "when_to_use": (
                    "Tasks involving schema changes, migrations, and backward compatibility. Best for data and API evolution."
                ),
                "multi_gen": "Yes — via GenerationRunner.",
            },
            "game": {
                "when_to_use": (
                    "Tournament-scored competitive scenarios with match execution. "
                    "Best for grid_ctf, othello, and other game-like environments."
                ),
                "multi_gen": "Yes — via GenerationRunner (native).",
            },
        }

    def to_markdown(self) -> str:
        lines = ["# Scenario Family Guide\n"]
        for family, info in self.families.items():
            lines.append(f"## {family}")
            lines.append(f"**When to use:** {info['when_to_use']}")
            lines.append(f"**Multi-generation:** {info['multi_gen']}\n")
        return "\n".join(lines)


GenerateFn = Callable[[str, int], str]
EvaluateFn = Callable[[str, int], AgentTaskGenerationEvaluation]


class AgentTaskEvolutionRunner:
    """Multi-generation runner for AgentTask scenarios with lesson accumulation."""

    def __init__(
        self,
        task_prompt: str,
        generate_fn: GenerateFn,
        evaluate_fn: EvaluateFn,
        initial_output: str = "",
        task_name: str = "agent_task",
        slot: FunctionSlot | None = None,
    ) -> None:
        self._task_prompt = task_prompt
        self._generate_fn = generate_fn
        self._evaluate_fn = evaluate_fn
        self._initial_output = initial_output
        self._task_name = task_name
        self._slot = slot

    def run_generation(
        self,
        state: AgentTaskGenerationState,
    ) -> AgentTaskGenerationState:
        """Run one generation: generate, evaluate, accumulate lessons, advance state."""
        prompt = build_enriched_prompt(
            task_prompt=self._task_prompt,
            playbook=state.playbook,
            generation=state.generation + 1,
            best_output=state.best_output,
            best_score=state.best_score,
            harness=self._slot.harness if self._slot else "",
        )

        if state.generation == 0 and self._initial_output:
            candidate_output = self._initial_output
        else:
            candidate_output = self._generate_fn(prompt, state.generation).strip()
            if not candidate_output:
                candidate_output = state.best_output

        if self._slot is not None:
            # Function-slot mode: evaluate the assembled harness+slot, but
            # carry only the small slot forward (no whole-program bloat).
            assembled = self._slot.assemble(candidate_output)
            evaluation = self._evaluate_fn(assembled, state.generation)
            evaluated_output = candidate_output
        else:
            evaluation = self._evaluate_fn(candidate_output, state.generation)
            evaluated_output = evaluation.output.strip() or candidate_output

        judge_result = AgentTaskResult(
            score=evaluation.score,
            reasoning=evaluation.reasoning,
            dimension_scores=evaluation.dimension_scores,
        )

        lesson = accumulate_lessons(judge_result, state.generation + 1, signal=evaluation.lesson_signal)
        new_playbook = state.playbook
        if lesson:
            new_playbook = (state.playbook + "\n" + lesson).strip() if state.playbook else lesson

        new_best_output = state.best_output
        new_best_score = state.best_score
        if not state.best_output or evaluation.score >= state.best_score:
            new_best_output = evaluated_output
            new_best_score = evaluation.score

        metadata = dict(state.metadata)
        generation_prompts = list(metadata.get("generation_prompts", []))
        generation_outputs = list(metadata.get("generation_outputs", []))
        generation_round_counts = list(metadata.get("generation_round_counts", []))
        met_threshold_history = list(metadata.get("met_threshold_history", []))

        generation_prompts.append(prompt)
        generation_outputs.append(evaluated_output)
        generation_round_counts.append(evaluation.round_count)
        met_threshold_history.append(evaluation.met_threshold)

        metadata["generation_prompts"] = generation_prompts
        metadata["generation_outputs"] = generation_outputs
        metadata["generation_round_counts"] = generation_round_counts
        metadata["met_threshold_history"] = met_threshold_history

        return AgentTaskGenerationState(
            generation=state.generation + 1,
            best_output=new_best_output,
            best_score=new_best_score,
            playbook=new_playbook,
            score_history=[*state.score_history, evaluation.score],
            lesson_history=[*state.lesson_history, lesson],
            metadata=metadata,
        )

    def run_with_state(
        self,
        num_generations: int = 10,
    ) -> tuple[AgentTaskTrajectory, AgentTaskGenerationState]:
        """Run multiple generations and return both trajectory and final state."""
        state = AgentTaskGenerationState(
            generation=0,
            best_output="",
            best_score=0.0,
            playbook="",
            score_history=[],
            lesson_history=[],
            metadata={},
        )

        for _ in range(num_generations):
            state = self.run_generation(state)

        trajectory = AgentTaskTrajectory(
            task_name=self._task_name,
            total_generations=num_generations,
            score_history=state.score_history,
            lessons_per_generation=[1 if lesson else 0 for lesson in state.lesson_history],
            cold_start_score=state.score_history[0] if state.score_history else 0.0,
            final_score=state.score_history[-1] if state.score_history else 0.0,
            improvement_delta=round(
                (state.score_history[-1] - state.score_history[0]) if state.score_history else 0.0,
                4,
            ),
            metadata={
                "best_output": state.best_output,
                "best_score": state.best_score,
                "playbook": state.playbook,
                "lesson_history": state.lesson_history,
                **state.metadata,
            },
        )
        return trajectory, state

    def run(self, num_generations: int = 10) -> AgentTaskTrajectory:
        """Run multiple generations and return a trajectory report."""
        trajectory, _ = self.run_with_state(num_generations)
        return trajectory

    def run_islands(
        self,
        num_islands: int = 4,
        num_generations: int = 10,
        migrate_every: int = 0,
    ) -> AgentTaskTrajectory:
        """Run ``num_islands`` parallel lineages, optionally migrating the
        global champion into laggards every ``migrate_every`` generations.

        Islands preserve diversity (each keeps its own playbook and lineage)
        while migration shares winners — the population analogue of the
        single-lineage :meth:`run`. ``migrate_every=0`` disables migration
        (pure parallel islands).
        """
        if num_islands < 1:
            raise ValueError(f"num_islands must be >= 1, got {num_islands}")
        states = [
            AgentTaskGenerationState(
                generation=0,
                best_output="",
                best_score=0.0,
                playbook="",
                score_history=[],
                lesson_history=[],
                metadata={},
            )
            for _ in range(num_islands)
        ]

        best_per_gen: list[float] = []
        for gen in range(num_generations):
            states = [self.run_generation(s) for s in states]
            best_per_gen.append(max(s.best_score for s in states))
            if migrate_every and (gen + 1) % migrate_every == 0:
                states = migrate_states(states)

        champion = max(states, key=lambda s: s.best_score)
        return AgentTaskTrajectory(
            task_name=self._task_name,
            total_generations=num_generations,
            score_history=best_per_gen,
            lessons_per_generation=[num_islands] * len(best_per_gen),
            cold_start_score=best_per_gen[0] if best_per_gen else 0.0,
            final_score=best_per_gen[-1] if best_per_gen else 0.0,
            improvement_delta=round((best_per_gen[-1] - best_per_gen[0]) if best_per_gen else 0.0, 4),
            metadata={
                "best_output": champion.best_output,
                "best_score": champion.best_score,
                "num_islands": num_islands,
                "playbook": champion.playbook,
            },
        )

File: ts/src/execution/agent-task-evolution.ts
/**
 * Multi-generation support for AgentTask scenarios.
 *
 * TypeScript port of `autocontext/execution/agent_task_evolution.py` (AC-281),
 * maintained at behavioral parity with the Python module. The framework's
 * native multi-generation loop for agent tasks: lesson accumulation,
 * best-tracking, and enriched prompts carried across generations.
 */

import { compactPromptComponent } from "../knowledge/semantic-compaction.js";
import type { AgentTaskResult } from "../types/index.js";

/** Cross-generation state for an agent task evolution run. */
export interface AgentTaskGenerationState {
  generation: number;
  bestOutput: string;
  bestScore: number;
  playbook: string;
  scoreHistory: number[];
  lessonHistory: string[];
  metadata: Record<string, unknown>;
}

/**
 * Structured, evaluator-provided guidance for the next generation.
 *
 * Domain-aware lesson accumulation: a deterministic evaluator usually knows
 * why a candidate plateaued and what to try next. Carrying that here lets
 * `accumulateLessons` write actionable playbook entries instead of only
 * score + dimension scores.
 */
export interface LessonSignal {
  hint?: string;
  plateau?: boolean;
  metrics?: Record<string, number>;
}

/** Evaluation result for one cross-generation candidate. */
export interface AgentTaskGenerationEvaluation {
  output: string;
  score: number;
  reasoning: string;
  dimensionScores?: Record<string, number>;
  roundCount?: number;
  metThreshold?: boolean;
  metadata?: Record<string, unknown>;
  lessonSignal?: LessonSignal;
}

/** Trajectory report for a multi-generation agent task run. */
export interface AgentTaskTrajectory {
  taskName: string;
  totalGenerations: number;
  scoreHistory: number[];
  lessonsPerGeneration: number[];
  coldStartScore: number;
  finalScore: number;
  improvementDelta: number;
  metadata: Record<string, unknown>;
}

/**
 * A fixed code harness with a small evolved slot (AC-776).
 *
 * Function-slot evolution mode keeps the evolved unit small: the runner
 * carries only the slot in state and in the enriched prompt (so prompts stay
 * compact), while evaluation runs the assembled harness + slot. This avoids
 * the whole-program-bloat failure mode where carrying a large generated
 * artifact in `bestOutput` ballooned every prompt.
 *
 * Convention: the slot is prepended to the harness so the harness can
 * reference names the slot defines.
 */
export class FunctionSlot {
  constructor(public readonly harness: string) {}

  /** Return the full runnable program: slot prepended to harness. */
  assemble(slot: string): string {
    return `${slot}\n\n${this.harness}`;
  }
}

export type GenerateFn = (prompt: string, generation: number) => string | Promise<string>;
export type EvaluateFn = (
  output: string,
  generation: number,
) => AgentTaskGenerationEvaluation | Promise<AgentTaskGenerationEvaluation>;

function fixed2(value: number): string {
  return value.toFixed(2);
}

function formatMetric(value: number): string {
  // Mirror Python's `{v:g}` general format (trims trailing zeros).
  return Number.isInteger(value) ? String(value) : String(value);
}

/**
 * Extract a structured lesson from judge feedback for the playbook.
 *
 * When the evaluator supplies a {@link LessonSignal}, its actionable guidance
 * (hint, plateau flag, metrics) is rendered alongside the score and dimension
 * scores so the playbook carries move-level direction.
 */
export function accumulateLessons(
  judgeResult: AgentTaskResult,
  generation: number,
  signal?: LessonSignal,
): string {
  const parts: string[] = [`Generation ${generation} (score: ${fixed2(judgeResult.score)}):`];

  if (judgeResult.reasoning) {
    parts.push(`  Feedback: ${judgeResult.reasoning}`);
  }

  const dims = judgeResult.dimensionScores ?? {};

  const weak = Object.entries(dims).filter(([, s]) => s < 0.7);
  if (weak.length > 0) {
    weak.sort((a, b) => a[1] - b[1]);
    const strs = weak.map(([d, s]) => `${d} (${fixed2(s)})`);
    parts.push(`  Weak dimensions: ${strs.join(", ")}`);
  }

  const strong = Object.entries(dims).filter(([, s]) => s >= 0.8);
  if (strong.length > 0) {
    strong.sort((a, b) => b[1] - a[1]);
    const strs = strong.map(([d, s]) => `${d} (${fixed2(s)})`);
    parts.push(`  Strong dimensions: ${strs.join(", ")}`);
  }

  if (!judgeResult.reasoning && weak.length === 0) {
    parts.push(`  Score: ${fixed2(judgeResult.score)}`);
  }

  if (signal) {
    if (signal.hint) {
      parts.push(`  Hint: ${signal.hint}`);
    }
    if (signal.plateau) {
      parts.push(
        "  Plateau detected — a structurally different approach is needed; " +
          "incremental tweaks are not advancing the score.",
      );
    }
    if (signal.metrics && Object.keys(signal.metrics).length > 0) {
      const metricStrs = Object.entries(signal.metrics)
        .sort((a, b) => a[0].localeCompare(b[0]))
        .map(([k, v]) => `${k}=${formatMetric(v)}`);
      parts.push(`  Metrics: ${metricStrs.join(", ")}`);
    }
  }

  return parts.join("\n");
}

/**
 * Enrich a task prompt with cross-generation context.
 *
 * In function-slot mode (`harness` provided), the fixed harness is shown once
 * as stable context so the model knows the contract it writes the slot
 * against. The evolved slot itself is carried via `bestOutput`.
 */
export function buildEnrichedPrompt(args: {
  taskPrompt: string;
  playbook: string;
  generation: number;
  bestOutput: string;
  bestScore: number;
  harness?: string;
}): string {
  const playbook = compactPromptComponent("agent_task_playbook", args.playbook);
  const bestOutput = compactPromptComponent("agent_task_best_output", args.bestOutput);
  const sections: string[] = [args.taskPrompt];

  if (args.harness) {
    sections.push(
      "\n\n## Fixed Harness (do not modify; you write only the slot)\n" + `${args.harness}`,
    );
  }

  if (playbook) {
    sections.push(
      `\n\n## Accumulated Lessons (Generation ${args.generation})\n` +
        `Previous best score: ${fixed2(args.bestScore)}\n\n` +
        `${playbook}`,
    );
  }

  if (bestOutput) {
    sections.push(
      `\n\n## Best Previous Output (score ${fixed2(args.bestScore)})\n` + `${bestOutput}`,
    );
  }

  if (playbook || bestOutput) {
    sections.push(
      "\n\nUse the accumulated lessons and previous best output as context. " +
        "Produce an improved version that addresses the identified weaknesses.",
    );
  }

  return sections.join("\n");
}

/**
 * Island migration: seed lagging islands with the global champion.
 *
 * Each island below the best score adopts the champion's best output and
 * score (winners propagate), but keeps its own playbook so accumulated
 * lessons stay diverse. The champion island (and any tied) are unchanged.
 */
export function migrateStates(states: AgentTaskGenerationState[]): AgentTaskGenerationState[] {
  if (states.length === 0) {
    return states;
  }
  let champion = states[0];
  for (const s of states) {
    if (s.bestScore > champion.bestScore) {
      champion = s;
    }
  }
  return states.map((s) =>
    s.bestScore < champion.bestScore
      ? { ...s, bestOutput: champion.bestOutput, bestScore: champion.bestScore }
      : s,
  );
}

/** Multi-generation runner for AgentTask scenarios with lesson accumulation. */
export class AgentTaskEvolutionRunner {
  private readonly taskPrompt: string;
  private readonly generateFn: GenerateFn;
  private readonly evaluateFn: EvaluateFn;
  private readonly initialOutput: string;
  private readonly taskName: string;
  private readonly slot: FunctionSlot | undefined;

  constructor(args: {
    taskPrompt: string;
    generateFn: GenerateFn;
    evaluateFn: EvaluateFn;
    initialOutput?: string;
    taskName?: string;
    slot?: FunctionSlot;
  }) {
    this.taskPrompt = args.taskPrompt;
    this.generateFn = args.generateFn;
    this.evaluateFn = args.evaluateFn;
    this.initialOutput = args.initialOutput ?? "";
    this.taskName = args.taskName ?? "agent_task";
    this.slot = args.slot;
  }

  async runGeneration(state: AgentTaskGenerationState): Promise<AgentTaskGenerationState> {
    const prompt = buildEnrichedPrompt({
      taskPrompt: this.taskPrompt,
      playbook: state.playbook,
      generation: state.generation + 1,
      bestOutput: state.bestOutput,
      bestScore: state.bestScore,
      harness: this.slot?.harness,
    });

    let candidateOutput: string;
    if (state.generation === 0 && this.initialOutput) {
      candidateOutput = this.initialOutput;
    } else {
      candidateOutput = (await this.generateFn(prompt, state.generation)).trim();
      if (!candidateOutput) {
        candidateOutput = state.bestOutput;
      }
    }

    let evaluation: AgentTaskGenerationEvaluation;
    let evaluatedOutput: string;
    if (this.slot !== undefined) {
      // Function-slot mode: evaluate the assembled harness+slot, but carry
      // only the small slot forward (no whole-program bloat).
      evaluation = await this.evaluateFn(this.slot.assemble(candidateOutput), state.generation);
      evaluatedOutput = candidateOutput;
    } else {
      evaluation = await this.evaluateFn(candidateOutput, state.generation);
      evaluatedOutput = evaluation.output.trim() || candidateOutput;
    }

    const judgeResult: AgentTaskResult = {
      score: evaluation.score,
      reasoning: evaluation.reasoning,
      dimensionScores: evaluation.dimensionScores ?? {},
      internalRetries: 0,
    };

    const lesson = accumulateLessons(judgeResult, state.generation + 1, evaluation.lessonSignal);
    let newPlaybook = state.playbook;
    if (lesson) {
      newPlaybook = state.playbook ? `${state.playbook}\n${lesson}`.trim() : lesson;
    }

    let newBestOutput = state.bestOutput;
    let newBestScore = state.bestScore;
    if (!state.bestOutput || evaluation.score >= state.bestScore) {
      newBestOutput = evaluatedOutput;
      newBestScore = evaluation.score;
    }

    const metadata: Record<string, unknown> = { ...state.metadata };
    const generationPrompts = [...((metadata.generationPrompts as string[]) ?? []), prompt];
    const generationOutputs = [
      ...((metadata.generationOutputs as string[]) ?? []),
      evaluatedOutput,
    ];
    const generationRoundCounts = [
      ...((metadata.generationRoundCounts as number[]) ?? []),
      evaluation.roundCount ?? 1,
    ];
    const metThresholdHistory = [
      ...((metadata.metThresholdHistory as boolean[]) ?? []),
      evaluation.metThreshold ?? false,
    ];
    metadata.generationPrompts = generationPrompts;
    metadata.generationOutputs = generationOutputs;
    metadata.generationRoundCounts = generationRoundCounts;
    metadata.metThresholdHistory = metThresholdHistory;

    return {
      generation: state.generation + 1,
      bestOutput: newBestOutput,
      bestScore: newBestScore,
      playbook: newPlaybook,
      scoreHistory: [...state.scoreHistory, evaluation.score],
      lessonHistory: [...state.lessonHistory, lesson],
      metadata,
    };
  }

  async runWithState(numGenerations = 10): Promise<{
    trajectory: AgentTaskTrajectory;
    state: AgentTaskGenerationState;
  }> {
    let state: AgentTaskGenerationState = {
      generation: 0,
      bestOutput: "",
      bestScore: 0,
      playbook: "",
      scoreHistory: [],
      lessonHistory: [],
      metadata: {},
    };

    for (let i = 0; i < numGenerations; i += 1) {
      state = await this.runGeneration(state);
    }

    const sh = state.scoreHistory;
    const delta = sh.length > 0 ? Math.round((sh[sh.length - 1] - sh[0]) * 1e4) / 1e4 : 0;
    const trajectory: AgentTaskTrajectory = {
      taskName: this.taskName,
      totalGenerations: numGenerations,
      scoreHistory: sh,
      lessonsPerGeneration: state.lessonHistory.map((l) => (l ? 1 : 0)),
      coldStartScore: sh.length > 0 ? sh[0] : 0,
      finalScore: sh.length > 0 ? sh[sh.length - 1] : 0,
      improvementDelta: delta,
      metadata: {
        bestOutput: state.bestOutput,
        bestScore: state.bestScore,
        playbook: state.playbook,
        lessonHistory: state.lessonHistory,
        ...state.metadata,
      },
    };
    return { trajectory, state };
  }

  async run(numGenerations = 10): Promise<AgentTaskTrajectory> {
    return (await this.runWithState(numGenerations)).trajectory;
  }

  /**
   * Run `numIslands` parallel lineages, optionally migrating the global
   * champion into laggards every `migrateEvery` generations. Islands
   * preserve diversity (each keeps its own playbook/lineage) while migration
   * shares winners. `migrateEvery = 0` disables migration.
   */
  async runIslands(
    args: {
      numIslands?: number;
      numGenerations?: number;
      migrateEvery?: number;
    } = {},
  ): Promise<AgentTaskTrajectory> {
    const numIslands = args.numIslands ?? 4;
    const numGenerations = args.numGenerations ?? 10;
    const migrateEvery = args.migrateEvery ?? 0;

    if (numIslands < 1) {
      throw new RangeError(`numIslands must be >= 1, got ${numIslands}`);
    }

    let states: AgentTaskGenerationState[] = Array.from({ length: numIslands }, () => ({
      generation: 0,
      bestOutput: "",
      bestScore: 0,
      playbook: "",
      scoreHistory: [],
      lessonHistory: [],
      metadata: {},
    }));

    const bestPerGen: number[] = [];
    for (let gen = 0; gen < numGenerations; gen += 1) {
      // Islands are independent within a generation — run them concurrently.
      states = await Promise.all(states.map((s) => this.runGeneration(s)));
      bestPerGen.push(Math.max(...states.map((s) => s.bestScore)));
      if (migrateEvery && (gen + 1) % migrateEvery === 0) {
        states = migrateStates(states);
      }
    }

    let champion = states[0];
    for (const s of states) {
      if (s.bestScore > champion.bestScore) {
        champion = s;
      }
    }
    const delta =
      bestPerGen.length > 0
        ? Math.round((bestPerGen[bestPerGen.length - 1] - bestPerGen[0]) * 1e4) / 1e4
        : 0;
    return {
      taskName: this.taskName,
      totalGenerations: numGenerations,
      scoreHistory: bestPerGen,
      lessonsPerGeneration: bestPerGen.map(() => numIslands),
      coldStartScore: bestPerGen.length > 0 ? bestPerGen[0] : 0,
      finalScore: bestPerGen.length > 0 ? bestPerGen[bestPerGen.length - 1] : 0,
      improvementDelta: delta,
      metadata: {
        bestOutput: champion.bestOutput,
        bestScore: champion.bestScore,
        numIslands,
        playbook: champion.playbook,
      },
    };
  }
}
